Simple S3 storage for sstables

xemul opened the pull request:
The PR adds sstables backend that works like existing local disk one but keeps Data and Index files as S3 objects. All other files including manipulations with them, remain intact. The objects on S3 get UUID names and those UUIDs are put as content into respective local files.

To play with:

1. Download and start minio
```
wget https://dl.minio.io/server/minio/release/linux-amd64/minio
chmod 755 minio
export MINIO_ROOT_USER=${root_user}
export MINIO_ROOT_PASSWORD=${root_pass}
mkdir -p ${root_directory}
minio server ${root_directory}
```

2. Configure minio CLI, create anonymous bucket
```
wget https://dl.min.io/client/mc/release/linux-amd64/mc
chmod 755 mc
mc config host rm local
mc config host add local http://127.0.0.1:9000 ${root_user} ${root_pass}
mc mb local/${bucket}
mc anonymous set public local/sstables
```

3. Start Scylla with object-storage feature enabled
``` scylla ... --experimental-features=keyspace-storage-options --workdir ${as_usual}```

4. Create KS with S3 storage
``` create keyspace ... storage = { 'type': 'S3', 'endpoint': '127.0.0.1:9000', 'bucket': 'sstables' };```

The S3 client has a logger named "s3", it's useful to use on with `trace` verbosity.

dorlaor commented:
Sweet!

On Sat, Jan 14, 2023 at 12:14 PM Pavel Emelyanov ***@***.***>
wrote:

> The PR adds sstables backend that works like existing local disk one but
> keeps Data and Index files as S3 objects. All other files including
> manipulations with them, remain intact. The objects on S3 get UUID names
> and those UUIDs are put as content into respective local files.
>
> To play with:
>
>    1. Download and start minio
>
> wget https://dl.minio.io/server/minio/release/linux-amd64/minio
> chmod 755 minio
> export MINIO_ROOT_USER=${root_user}
> export MINIO_ROOT_PASSWORD=${root_pass}
> mkdir -p ${root_directory}
> minio server ${root_directory}
>
>
>    1. Configure minio CLI, create anonymous bucket
>
> wget https://dl.min.io/client/mc/release/linux-amd64/mc
> chmod 755 mc
> mc config host rm local
> mc config host add local http://127.0.0.1:9000 ${root_user} ${root_pass}
> mc mb local/${bucket}
> mc anonymous set public local/sstables
>
>
>    1.
>
>    Start Scylla with object-storage feature enabled
>    scylla ... --experimental-features=keyspace-storage-options --workdir
>    ${as_usual}
>    2.
>
>    Create KS with S3 storage
>    create keyspace ... storage = { 'type': 'S3', 'endpoint': '
>    127.0.0.1:9000', 'bucket': 'sstables' };
>
> The S3 client has a logger named "s3", it's useful to use on with trace
> verbosity.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/scylladb/scylladb/pull/12523
> Commit Summary
>
>    - b9719d1
>    <https://github.com/scylladb/scylladb/pull/12523/commits/b9719d1859790f9317c346d633fd32a73a3e6099>
>    Update seastar with http client
>    - 53b6669
>    <https://github.com/scylladb/scylladb/pull/12523/commits/53b6669e6f5538ac5a5e6e84dfdb24a09e91edb4>
>    memory_data_sink: Add move ctor
>    - 924e27b
>    <https://github.com/scylladb/scylladb/pull/12523/commits/924e27b9954c12139a12fff7aa2676071d9e0afd>
>    utils: Add S3 client with basic ops
>    - fdc9adc
>    <https://github.com/scylladb/scylladb/pull/12523/commits/fdc9adcc1f1b248925541e605b10e5ed90ddd52e>
>    utils: Add S3 data sink for multipart upload
>    - fcbad32
>    <https://github.com/scylladb/scylladb/pull/12523/commits/fcbad32bf5cc7de239710fd1c7eec417add4c3de>
>    utils: Add S3 readable file impl for random reads
>    - 2b57404
>    <https://github.com/scylladb/scylladb/pull/12523/commits/2b57404f33e247d004f81df62c36aba2a8ea330f>
>    test: Fixie for test sstable chdir
>    - 1ea31f5
>    <https://github.com/scylladb/scylladb/pull/12523/commits/1ea31f53d0e5fcb08d0ef1dcc5dc394dd8757c49>
>    sstable: Make storage an API
>    - 55def37
>    <https://github.com/scylladb/scylladb/pull/12523/commits/55def372522294067ec7405b281560d600243b07>
>    sstable/writer: Shuffle writer::init_file_writers()
>    - a3d8a04
>    <https://github.com/scylladb/scylladb/pull/12523/commits/a3d8a04cb1f8c125b41513728ae3e5131e4d88ac>
>    sstable, storage: Virtualize data sink making
>    - 8bfe431
>    <https://github.com/scylladb/scylladb/pull/12523/commits/8bfe431041b3050c69a0efef10df8472d98a4781>
>    keyspace_metadata: Make storage options lw-shared-ptr
>    - 0b18471
>    <https://github.com/scylladb/scylladb/pull/12523/commits/0b18471403fac37222ab6c59b3f0ad6c9251106d>
>    table: Keep storage options lw-shared-ptr
>    - 1d7a05a
>    <https://github.com/scylladb/scylladb/pull/12523/commits/1d7a05a8289849cdab3956185270531cc4207292>
>    sstable: Make storage instance based on storage options
>    - 7c207ad
>    <https://github.com/scylladb/scylladb/pull/12523/commits/7c207ade94aba6784a4f2dd24765df02db28aa0b>
>    sstables: Add S3 data and index storage implementation
>
> File Changes
>
> (28 files <https://github.com/scylladb/scylladb/pull/12523/files>)
>
>    - *M* configure.py
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-4d5f3192809ec1b9add6b33007e0c50031ad9a0a2f3f55a481b506468824db2c>
>    (2)
>    - *M* data_dictionary/data_dictionary.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-ede0d02db2aaa3bd5726f165ae39a6442de10c9e654fe7587564431d394c52de>
>    (2)
>    - *M* data_dictionary/keyspace_metadata.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-9ca5263f79b584831826132b9eeafa8f17de8865e248c10b4ffc6ba67907d614>
>    (6)
>    - *M* data_dictionary/storage_options.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-305d81570c85d930414ef0998053a3b9cb50bf4aede403f6a8a255180f02fb88>
>    (2)
>    - *M* replica/database.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-5f674f2f5beecae883403216eef63b543f1bb5aaa47c95089468386d2f4b9fdf>
>    (4)
>    - *M* replica/database.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-948bf31fedfd133709ed39b734ac8ccf1fd0510378a6c10e0d9c50ba03d298f2>
>    (13)
>    - *M* replica/table.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-970063d9f766ba24975f1955091d6bd0b071932333c47936fe2b4ff62032c580>
>    (7)
>    - *M* seastar
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-38873f10a07ef58586b6fb3ec38658a256a2f1f3882c88ea0bcbfb9d1f91a769>
>    (2)
>    - *M* sstables/mx/writer.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-5ab68b92b6216a1d090f36634983103ce1ae4a4e8b42784cdd38ebbebbd1ab3c>
>    (16)
>    - *M* sstables/sstable_directory.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-e79b7a00f84da94b4eb9f541180527298f6548505fcf8c00bc5a19b529dff004>
>    (10)
>    - *M* sstables/sstable_directory.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-dfb4fcc4e73f026e553890801aa355a7ee4d6853ff007e2e310438904d91050a>
>    (3)
>    - *M* sstables/sstables.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-e77ac6a4425e9a7eec3d461a5896cc62a6fcca38b02ad855e047f9baace8b5b3>
>    (166)
>    - *M* sstables/sstables.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-ffd7a1cabc77a7c96ad97c1d21c0fb84746b53b97167774f8d9e577379fecd7a>
>    (46)
>    - *M* sstables/sstables_manager.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-409f6359b6f62b8d3842fb6fafbb674706d29158a4f033880a550e82d2589741>
>    (3)
>    - *M* sstables/sstables_manager.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-3df6baf65a7194304b8e2e13afffd6573deb61ec30288acc069861b12011db51>
>    (1)
>    - *M* test/boost/mutation_test.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-55d8baee30e4b9dddd1b423ae73b117bdd25dbdd7ec27c1765dcb97b66ec6c4a>
>    (3)
>    - *M* test/boost/sstable_3_x_test.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-eccc4c02001501042d01bf4c00a8789c94efb483b47dd43bbf9f681da58d00b2>
>    (2)
>    - *M* test/boost/sstable_compaction_test.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-5a3efb17a5da346667e9aee09684d3ed2e42c8320ae103babd32f666659ed38a>
>    (2)
>    - *M* test/boost/sstable_directory_test.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-13b632e8a9eeb2ae4e05c6d8b5c129c084ac184ddd6a62c6f451ebe82681c393>
>    (3)
>    - *M* test/lib/sstable_test_env.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-b191be387c602a17f532b83d7d30bd48db79efd15e81c6dadc00daa20950ca69>
>    (4)
>    - *M* test/lib/sstable_utils.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-29f8c9c234fd76745e53e41b2aba797fddee395c246bb1ff3b6bc2847e6b04be>
>    (6)
>    - *M* test/lib/test_services.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-bcd968299c2c01fcb45fa2b939a45fca583790fb984650932bc04417107a77c7>
>    (2)
>    - *A* test/manual/s3_test.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-1419503133a1b7d6792e3bbca421cb6f559d2c3388e76b6a14b063994873c683>
>    (164)
>    - *M* test/perf/perf_sstable.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-2d497484344088308e3e3d860719261070e91aca064b9291a21c515ce3bde33b>
>    (2)
>    - *M* tools/scylla-sstable.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-4cd1a5351a4e8c9836e4cdf6eaa67498e53d365fe9ab76dcae4baa9e6ac4cdc0>
>    (6)
>    - *M* utils/memory_data_sink.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-e357f3c67a002983f725701c4cca4820670061be7380331932ee5486ce0e57da>
>    (8)
>    - *A* utils/s3/client.cc
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-28c1d03a5facd99f25f466e6bae94c0fab2bb8845c0538dd489db7832505e884>
>    (380)
>    - *A* utils/s3/client.hh
>    <https://github.com/scylladb/scylladb/pull/12523/files#diff-7a905c59f797a5ce1418dd124970d3b74e1464da6052fddb4a6c7b9fcd9e0803>
>    (47)
>
> Patch Links:
>
>    - https://github.com/scylladb/scylladb/pull/12523.patch
>    - https://github.com/scylladb/scylladb/pull/12523.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/12523>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANHURI7NJCVAGVBU5OZC3DWSJ4BZANCNFSM6AAAAAAT3GGADA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

avikivity commented:
> The PR adds sstables backend that works like existing local disk one but keeps Data and Index files as S3 objects. All other files including manipulations with them, remain intact. The objects on S3 get UUID names and those UUIDs are put as content into respective local files.
> 
> To play with:
> 
> 1. Download and start minio
> 
> ```
> wget https://dl.minio.io/server/minio/release/linux-amd64/minio
> chmod 755 minio
> export MINIO_ROOT_USER=${root_user}
> export MINIO_ROOT_PASSWORD=${root_pass}
> mkdir -p ${root_directory}
> minio server ${root_directory}
> ```

We'll want to add minio to the frozen toolchain for testing. Please send a patch (separately) adding it to install-dependencies.sh.

I'd like most testing to work using regular directories (shared among nodes) for simplicity, but we'll need some S3 testing.

> 
> 2. Configure minio CLI, create anonymous bucket
> 
> ```
> wget https://dl.min.io/client/mc/release/linux-amd64/mc
> chmod 755 mc
> mc config host rm local
> mc config host add local http://127.0.0.1:9000 ${root_user} ${root_pass}
> mc mb local/sstables
> mc anonymous set public local/sstables
> ```
> 
> 3. Start Scylla with object-storage feature enabled
>    ` scylla ... --experimental-features=keyspace-storage-options --workdir ${as_usual}`
> 4. Create KS with S3 storage
>    ` create keyspace ... storage = { 'type': 'S3', 'endpoint': '127.0.0.1:9000', 'bucket': 'sstables' };`
> 
> The S3 client has a logger named "s3", it's useful to use on with `trace` verbosity.

avikivity commented:
re the seastar submodule update, it currently breaks some random dtest, see https://github.com/scylladb/scylladb/pull/12509

xemul commented:
> re the seastar submodule update, it currently breaks some random dtest, see #12509

That's OK, seastar updating commit here includes custom branch with not yet merged PRs

raphaelsc commented on utils/s3/client.cc:
```diff
@@ -0,0 +1,105 @@
+ */
+
+#include <rapidxml.h>
+#include <seastar/core/coroutine.hh>
```
rapidxml must be added to the dependency list

avikivity replied:
From the website:

**Last Updated**
2013-04-26

While that's my birthday, maybe we should look for something more recent. On the other hand, if it works, it works.

xemul replied:
No incidents reported since 2013 :sunglasses:

raphaelsc commented on test/manual/s3_test.cc:
```diff
@@ -0,0 +1,63 @@
+            BOOST_REQUIRE(false);
+        } catch (const std::exception& e) {
+            fmt::print(" -> {}\n", e.what());
+        }
```
alternatively, we can do something like:

BOOST_REQUIRE_EXCEPTION(cln->get_object_size(name).get(), std::runtime_error, exception_predicate::message_contains("404 Not Found"))

so we assert that s3 server replies with the expected error.

raphaelsc commented on utils/s3/client.hh:
```diff
@@ -0,0 +1,41 @@
+    static shared_ptr<client> make(ipv4_addr addr);
+
+    future<uint64_t> get_object_size(sstring object_name);
+    future<temporary_buffer<char>> get_object_contiguous(sstring object_name, std::optional<range> range = {});
```
sounds good we'll have only one interface for retrieving a range of an object into a contiguous buffer. sstable layer must be responsible anyway for not reading a large range in a single iteration.

raphaelsc commented on utils/s3/client.hh:
```diff
@@ -0,0 +1,41 @@
+    explicit client(ipv4_addr addr, private_tag);
+    static shared_ptr<client> make(ipv4_addr addr);
+
+    future<uint64_t> get_object_size(sstring object_name);
```
can be made const.

xemul replied:
It looks like it cannot :( It calls `_http.make_request` inside which is not const and compiler complains. The latter, in turn, may start creating new connection to the server thus modifying the "connections pool" and cannot be marked `const`

avikivity replied:
We can mark _http mutable, if we think it's important. _http is like a pointer to the remote server.

raphaelsc commented on test/manual/s3_test.cc:
```diff
@@ -61,3 +61,58 @@ SEASTAR_TEST_CASE(test_client_put_get_object) {
         cln->close().get();
     });
```
"Putting a large object into S3 using plain PUT is bad choice -- one need
to collect the whole object in memory, then send it as a content-length
request with plain body. "

not always. Some sstable components such as Statistics.DB and TOC.txt will always be small (smaller than 8k), therefore they're good candidates for s3 PUT interface.

raphaelsc replied:
Ah you said large. Forget it :-)

raphaelsc commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+    static constexpr size_t minimum_part_size = 5 << 20;
+    static constexpr int flush_concurrency = 3;
+
```
this seems like write behind for files. of course, network concurrency cannot be treated the same as disk one, but perhaps make it configurable?

xemul replied:
Will do, yes

xemul replied:
#13021

avikivity commented on configure.py:
```diff
@@ -1030,6 +1031,7 @@ def find_headers(repodir, excluded_dirs):
                 'service/broadcast_tables/experimental/lang.cc',
                 'tasks/task_manager.cc',
                 'rust/wasmtime_bindings/src/lib.rs',
+                'utils/s3/client.cc',
```
Please put it next to the other utils.

xemul replied:
"Other utils" are somewhat dispersed throughout this list already :confused:

avikivity replied:
Well, next to some other group, let's not increase the dispersion.

avikivity commented on test/manual/s3_test.cc:
```diff
@@ -0,0 +1,63 @@
+ */
+
+static const ipv4_addr s3_server("127.0.0.1", 9000);
+
```
Let's move it away from test/manual and run minio. Integration into test.py will be annoying, but maybe it's possible to use boost.test fixures to launch minio directly from the test.

xemul replied:
Sure

avikivity commented on utils/s3/client.cc:
```diff
@@ -0,0 +1,105 @@
+    auto req = http::request::make("GET", _host, object_name);
+    http::reply::status_type expected = http::reply::status_type::ok;
+    if (range) {
+        req._headers["Range"] = format("bytes={}-{}", range->off, range->off + range->len - 1);
```
I feel pain seeing an I/O interface specify byte ranges in decimal :face_with_head_bandage:

xemul replied:
So do I. I think there's already enough pain about S3 API for a "how not to design object storage API" blog post...

avikivity commented on utils/s3/client.cc:
```diff
@@ -0,0 +1,105 @@
+    std::optional<temporary_buffer<char>> ret;
+    co_await _http.make_request(std::move(req), [&off, &ret, &object_name] (http::reply rep, input_stream<char>&& in_) mutable -> future<> {
+        auto in = std::move(in_);
+        ret = temporary_buffer<char>(rep.content_length);
```
We will want to add (later) a customization point so we can wrap this with a make_tracked_temporary_buffer(). We may also want to adjust the sstable reader to be able to consume buffers as they are read in, rather than wait+copy. The latter change is very large for dubious gains, so only after we gain some experience with it.

avikivity replied:
Please create an issue and link it into the main object storage tracker so we don't forget.

xemul replied:
#13306

avikivity commented on utils/s3/client.cc:
```diff
@@ -0,0 +1,105 @@
+    s3l.trace("PUT {}", object_name);
+    auto req = http::request::make("PUT", _host, object_name);
+    auto len = buf.size();
+    req.write_body("bin", len, [buf = std::move(buf)] (output_stream<char>&& out_) mutable -> future<> {
```
Need coroutine::lambda?

avikivity replied:
And then you can capture buf with `[&]`.

xemul replied:
The `request::write_body`'s argument is `noncopyable_function`. Doesn't it handle the lambda's lifecycle correctly itself?

avikivity commented on utils/s3/client.hh:
```diff
@@ -0,0 +1,41 @@
+    future<std::vector<temporary_buffer<char>>> get_object(sstring object_name, std::optional<std::pair<uint64_t, size_t>> range);
+
+public:
+    explicit client(ipv4_addr addr, private_tag);
```
Will socket_address work?

xemul replied:
Should work, yes. Will fix

asias replied:
Does seastar http code work with ipv6? If it works, s3::client should work with ipv6 as well, right?

xemul replied:
Seastar http code uses connected_socket whatever it is, so it would work with IPv6 too. I made this change in http, but apparently forgot to do the same with S3

avikivity commented on utils/s3/client.hh:
```diff
@@ -0,0 +1,41 @@
+    future<temporary_buffer<char>> get_object_contiguous(sstring object_name, std::optional<range> range = {});
+    future<> put_object(sstring object_name, temporary_buffer<char> buf);
+    future<> delete_object(sstring object_name);
+
```
Will we need an abort() here? Usually disk I/O is expected to complete in finite time, but network partition can cause us to time out here. On the other hand, if we're partitioned, nobody cares if we hang during shutdown or not.

I think we can avoid adding abort(), and if it turns out to be a problem, add tcp keepalive.

avikivity commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+    shared_ptr<client> _client;
+    http::experimental::client& _http;
+    sstring _object_name;
+    memory_data_sink_buffers _bufs;
```
Nice reuse

avikivity commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+        doc.parse<0>(body.data());
+    } catch (const rapidxml::parse_error& e) {
+        s3l.warn("cannot parse initiate multipart upload response: {}", e.what());
+        return "";
```
Shouldn't we throw? Or can we recover?

xemul replied:
The caller checks for the returned value to be empty and resolves with relevant exceptional future

avikivity replied:
Ok. Please document it whereever we swallow an exception, it screams "there is an error handling bug here".

avikivity commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+
+sstring parse_multipart_upload_id(sstring& body) {
+    rapidxml::xml_document<> doc;
+    std::exception_ptr ex;
```
Not used?

avikivity commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+            return "";
+        }
+
+        body += format("<Part><ETag>{}</ETag><PartNumber>{}</PartNumber></Part>", etag, nr++);
```
sstring += is quadratic. Better to use std::string.

10,000 parts * (say) 50 bytes = 500k. It's a little uncomfortable.

avikivity replied:
(or std::ostringstream)

avikivity commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+    req._headers["Content-Length"] = format("{}", bufs.size());
+    req.query_parameters["partNumber"] = format("{}", part_number);
+    req.query_parameters["uploadId"] = _upload_id;
+    req.write_body("bin", bufs.size(), [this, part_number, bufs = std::move(bufs)] (output_stream<char>&& out_) mutable -> future<> {
```
Same concern about lambda coroutines.

avikivity commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+    });
+
+    auto units = co_await get_units(_flush_sem, 1);
+    (void)_http.make_request(std::move(req), [this, part_number] (http::reply rep, input_stream<char>&& in_) mutable -> future<> {
```
Need comment justifying ignored future and explaining error handling.

asias replied:
I am wondering why we do not wait for make_request in client::upload_sink::upload_part in the first place?

xemul replied:
To issue upload parts in parallel. It will wait in `get_units()` if runs out of parallelizm

avikivity commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+        _part_etags[part_number] = std::move(etag);
+        return make_ready_future<>();
+    }).handle_exception([this, part_number] (auto ex) {
+        s3l.warn("couldn't upload part {}: {} (upload id {})", part_number, ex, _upload_id);
```
(and need to fix error handling)

xemul replied:
Can you elaborate? If the part uploading fails there appears a "hole" in the `_part_etags`, and at the end there can be two outcomes:
* flush() is called, it notices the hole and aborts the upload
* close() is called w/o flush, it notices the upload is not flushed and aborts the upload as well

avikivity replied:
If it works, that's fine, but a swallowed exception needs a comment explaining why it's okay.

It would also be good to test failures, though maybe hard. Maybe setup minio with a quota and upload something too large?

xemul replied:
S3 protocol is very fragile and it's easy to trigger server error by (unwillingly) violating any of 132645 corner cases. I've a patch that adds random delays to upload-parts that I used to test upload-finish synchronization. I'll think towards adding "error injection" too

avikivity commented on utils/s3/client.cc:
```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+    return data_sink(std::make_unique<upload_sink>(shared_from_this(), std::move(object_name)));
+}
+
 future<> client::close() {
```
10,000 parts * 5MB = 50GB. Too small for comfort.

If we multiply minimum_part_size by a factor k each time, we can make it arbitrarily large.

Looks like the sum is first_part*(1-k^n)/(1-k).

My math isn't up to solving it, but experimenting shows k=1.0005 is more than enough. The last part ends up 750GB though.

avikivity replied:
1.00045 gets us to 1TB with with "just" 445MB last part.

raphaelsc replied:
This is very important, specially because we don't split large partitions yet. I have seen >200G large partition in a single sstable. I think with today's implementation, the upload will just fail as S3 will reject another part once the 10k limit is reached.

@avikivity I think you were referring to UploadPart instead, as UploadPartCopy requires using existing objects as data source.

avikivity replied:
I was referring to UploadPartCopy. The first level uses 10,000 UploadPart * 5MB = 50GB to create several 50GB sub-objects. The second level uses UploadPartCopy to splice multiple 50GB sub-objects into a single 182TB sstable containing a single partition, and winning the large partition trophy for 2023.

avikivity replied:
(but, we needn't do it just now)

raphaelsc replied:
Ah I was focused on your first proposal of gradually increasing part (chunk) size. It's not good (throughput wise) that we have to do a second pass on the uploaded data, but I think you prefer it to avoid stressing node's memory when the input data is very very large.

avikivity replied:
It's S3 that has to do this second pass. It's not nice but it allows keeping memory usage at 5MB.

xemul replied:
Is there a way to pre-estimate the final object size somehow? If the object is going to be below 50Gb, then "default" uploading would look like

```
CreateUpload(temp)
UploadPart * N
CompleteMultipartUpload(temp)
CreateUpload(obj)
UploadPartCopy(from temp)
CompleteMultipartUpload(obj)
```

while it could be 2x times shorter

xemul replied:
If we assume that it's under 50GB we still need to send some value in the ContentLength header, and if the real  amount of data turns out to be less than that we're in trouble :(

raphaelsc replied:
If we know memtable or compaction will produce output which is smaller than 50G, then s3 client can task the fast path and it seems to me it can avoid UploadPartCopy altogether. As for compaction, it can definitely provide a deterministic response that it won't generate output smaller than 50G. Perhaps the only edge case is if user disabled compression and the output is larger than input, but we can detect it using compression ratio statistics in the SSTable and looking at the schema (if compression has been disabled).

As for ContentLength header, it doesn't really matter because we still buffer the output in 5MB chunks and flush it either after the buffer is full or operation, e.g. compaction, has exhausted the input data.

avikivity commented on utils/s3/client.cc:
```diff
@@ -16,6 +16,20 @@
 
+namespace utils {
+
+// FIXME -- the very same thing exists in seastar
```
That's fine. Seastar isn't a library for manipulating iovecs.

raphaelsc replied:
Couldn't resist to make it a one-liner: 

return boost::accumulate(iov | boost::adaptors::transformed(std::mem_fn(&iovec::iov_len)), size_t(0));

avikivity commented on data_dictionary/storage_options.hh:
```diff
@@ -29,6 +29,8 @@ struct storage_options {
 
     storage_options() = default;
+    storage_options(storage_options&&) = default;
+    storage_options(const storage_options&) = default;
```
Seems unnecessary, why are those needed? The compiler should synthesize them without help.

avikivity commented on replica/database.cc:
```diff
@@ -943,9 +943,9 @@ void database::add_column_family(keyspace& ks, schema_ptr schema, column_family:
                 ? *_schema_commitlog
                 : *_commitlog;
-        cf = make_lw_shared<column_family>(schema, std::move(cfg), cl, _compaction_manager, sst_manager, *_cl_stats, _row_cache_tracker);
+        cf = make_lw_shared<column_family>(schema, std::move(cfg), ks.metadata()->get_storage_options_ptr(), cl, _compaction_manager, sst_manager, *_cl_stats, _row_cache_tracker);
```
We might also just copy the thing.

avikivity commented on sstables/sstables.cc:
```diff
@@ -3218,14 +3219,100 @@ mutation_source sstable::as_mutation_source() {
+future<file> sstable::filesystem_with_s3_data_storage::open_component(const sstable& sst, component_type type, open_flags flags, file_open_options options, bool check_integrity) {
+    auto f = co_await filesystem_storage::open_component(sst, type, flags, options, check_integrity);
+
+    if (type == component_type::Data || type == component_type::Index) {
```
Why not extend to all the other components?

xemul replied:
Mainly because it needs more care around `make_component_file_writer()` that's used for other components. I'll include it in the next version

avikivity commented:
Looks very good, and easy to review too.

We might have a separate abstraction layer for sstable I/O instead of file::read, but what's in the series is good too.

asias commented on utils/s3/client.hh:
```diff
@@ -0,0 +1,47 @@
+    http::experimental::client _http;
+
+    struct private_tag {};
+    future<std::vector<temporary_buffer<char>>> get_object(sstring object_name, std::optional<std::pair<uint64_t, size_t>> range);
```
std::optional<range> range?

xemul replied:
Yikes, it's a trash forward-declaration, need to remove one

asias commented on utils/s3/client.hh:
```diff
@@ -0,0 +1,47 @@
+
+    future<uint64_t> get_object_size(sstring object_name);
+    future<temporary_buffer<char>> get_object_contiguous(sstring object_name, std::optional<range> range = {});
+    future<> put_object(sstring object_name, temporary_buffer<char> buf);
```
I see put_object is used only in tests. We do not support writing sstables to s3 yet in this version?

xemul replied:
Writing sstables is implemented in the next patch that adds uploading sink. PUT-ing an object is no-go for sstables as it needs knowing exact sstable size in advance

asias commented on sstables/sstables.cc:
```diff
@@ -3176,7 +3219,106 @@ mutation_source sstable::as_mutation_source() {
+future<data_sink> sstable::filesystem_with_s3_data_storage::make_component_sink(sstable& sst, component_type type, io_priority_class pc) {
+    assert(type == component_type::Data || type == component_type::Index);
+    sstring uuid = co_await get_object_uuid(sst, type);
+    co_return _client->make_upload_sink("/" + _bucket + "/" + uuid);
```
Ah, so here, we create the s3 upload sink and upload sstable files to s3.

asias commented:
After

```
create keyspace ... storage = { 'type': 'S3', 'endpoint': '127.0.0.1:9000', 'bucket': 'sstables' };
```

what will happen if user alters the keyspace back to local stroage?

As a starter, I guess we need to at least reject such operation.

xemul commented:
update:
- rebased
- added local ownership table (not yet replicated)
- keep all component files as S3 objects
- fixed some (all the simple ones) comments from the previous review

xemul commented:
UPD:
- added more pre-requisite PRs (first commit to review is 076b277fda0af0dc6c46b9748551a4f3a82b8db0)
- changed the ownership table schema
- changed the object paths
- keep TOC file as object
- ownership table it considered on boot, sstables are re-populated

The new ownership table schema (2f7281d5c3e0777cbb400f068a367366d914e31a) is
```
    CREATE TABLE system.sstables_registry (
        location text,
        generation bigint,
        format text,
        status text,
        uuid uuid,
        version text,
        PRIMARY KEY (location, generation)
    )
```

The new object paths are like

```
s3://test_bucket/d0a743b0-ad38-11ed-85b5-39b6b0998182/Data.db
```

TOC file is needed because `sstable::load()` calls `read_toc()` and cannot survive without it (yet?)

xemul commented:
With the :point_up: update it's finally minimally usable, so sstables are indeed persistent and outlive scylla instance :)
So the ownership table and S3-side objects layout are about to be settled in their current form

xemul commented:
update:
- rebased
- fixed multipart-upload completion xml dumping
- other review comments fixed
- moved s3 client test into boost/ suite (now it's automatic)

xemul commented:
updated:
- ready for merging (all required PRs are in master)
- added simple object_storage test to validate simple start-populate-restart functionality

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4917/

xemul commented:
> https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4917/

```
[2023-03-13T17:00:56.915Z] utils/s3/client.cc:9:10: fatal error: 'rapidxml.h' file not found
[2023-03-13T17:00:56.915Z] #include <rapidxml.h>
[2023-03-13T17:00:56.915Z]          ^~~~~~~~~~~~
[2023-03-13T17:00:56.915Z] 1 error generated.
```

:flushed:

denesb commented:
I will review this today or tomorrow.

raphaelsc commented on db/system_keyspace.cc:
```diff
@@ -964,6 +966,24 @@ schema_ptr system_keyspace::broadcast_kv_store() {
+        return schema_builder(NAME, SSTABLES_REGISTRY, id)
+            .with_column("location", utf8_type, column_kind::partition_key)
+            .with_column("generation", long_type, column_kind::clustering_key)
+            .with_column("uuid", uuid_type)
```
why both generation and uuid?

with uuid based generation, it seems having both generation and uuid is redundant.

additionally, we can have collision with 64-bit generation being used as clustering key.

even if we use uuid, using generation_type::value() means we're losing half of the info from the 128-bit uuid.

I suggest using uuid_type for generation, and refusing integer based generation with s3 storage.

xemul replied:
> why both generation and uuid?

generation because it's still there and it's needed by the rest of sstables code
uuid is here because I need a unique identifier for object on the s3 side

> with uuid based generation, it seems having both generation and uuid is redundant.

Of course. But as long as it's not yet there, we have to stick with both

> additionally, we can have collision with 64-bit generation being used as clustering key.

Hm... Nowadays 64-bit generation is used as a "unique" identifier for files on /var/lib/data directory, so having a conflict as clustering key means we can have conflict on filesystem too

> even if we use uuid, using generation_type::value() means we're losing half of the info from the 128-bit uuid.
> 
> I suggest using uuid_type for generation, and refusing integer based generation with s3 storage.

Definitely. Once uuid generations are in we'll do it

xemul replied:
The uuid field is 100% auxiliary here. It's only purpose it to provide S3-side unique object name. Once uuid-generation gets merged this field will be just removed from the table. I'm all for simplifying this part right now, but I don't see a way to shoehorn generation into a uuid just because generation is the clustering key, and it's essential to be such, and uuid is regular column which cannot be part of the clustering key, because it's the value the code needs to resolve with `location:generation` pair at hand.

Maybe an option to lower the WTF level around this thing is not to call it uuid and not to generate it as uuid? But use smth like random::get_random_string() instead? This would more-or-less work for "s3-side uniqueness" temporarily.

raphaelsc replied:
> > why both generation and uuid?
> 
> generation because it's still there and it's needed by the rest of sstables code uuid is here because I need a unique identifier for object on the s3 side
> 
> > with uuid based generation, it seems having both generation and uuid is redundant.
> 
> Of course. But as long as it's not yet there, we have to stick with both
> 
> > additionally, we can have collision with 64-bit generation being used as clustering key.
> 
> Hm... Nowadays 64-bit generation is used as a "unique" identifier for files on /var/lib/data directory, so having a conflict as clustering key means we can have conflict on filesystem too
> 

I was thinking ahead to shared storage, where 64-bit generation can collide across nodes in the same cluster.

So the idea is that we can modify the schema at time we want as the feature is not GA yet. Got it. thanks for the explanation.

raphaelsc replied:
Ah skipped this in the cover letter: "maintains the system.sstables_registry table to keep track of what objects belong to local node and their names." :-) sorry for the noise.

raphaelsc commented on db/system_keyspace.cc:
```diff
@@ -3421,6 +3444,54 @@ future<> system_keyspace::save_group0_upgrade_state(service::group0_upgrade_stat
 }
 
+future<> system_keyspace::sstables_reg_create_entry(sstring location, utils::UUID uuid, sstring status, sstables::entry_descriptor desc) {
+    static const auto req = format("INSERT INTO system.{} (location, generation, uuid, status, version, format) VALUES (?, ?, ?, ?, ?, ?)", SSTABLES_REGISTRY);
```
it would be nice to expand reg into registry. I had to jump here in order to figure out what reg stands for. but I understand it can become too long. Maybe we can remove the _entry suffix from function names.

denesb reviewed (commented):
It would be nice to get a bird's eye view of how this work in the cover letter. I did find out eventually, but I had to get to half-way into the series to learn how sstables ownership is stored.

I think the refactoring patches could be upstreamed separately while the meat of this PR is under review/discussion. They are distracting and will bit-rot fast.

Question for the future: AFAIK, one of the main goals of object storage is to allow nodes to store storage independently of compute. So far it seems to me that remote sstables are treated the same way as local ones, i.e. some of their component is kept in memory at all times. This means that the node is still bound to the usual memory:disk ration, regardless of where the sstables are located. What are the plans to mitigate this?

xemul commented:
> It would be nice to get a bird's eye view of how this work in the cover letter. I did find out eventually, but I had to get to half-way into the series to learn how sstables ownership is stored.

Will add more text to the PR description

> I think the refactoring patches could be upstreamed separately while the meat of this PR is under review/discussion. They are distracting and will bit-rot fast.

Well, most of the _real_ refactoring (and some fixes) had been merged already (#12881, #12847, #12843, #12771, #12648, etc (and a whole lot of test/ refactoring)). I'll check what can I squeeze more from this PR

> Question for the future: AFAIK, one of the main goals of object storage is to allow nodes to scale storage independently of compute. So far it seems to me that remote sstables are treated the same way as local ones, i.e. some of their component is kept in memory at all times. This means that the node is still bound to the usual memory:disk ratio, regardless of where the sstables are located. What are the plans to mitigate this?

Ah, thank you for reminding! I asked this very question several weeks ago on object storage weekly meeting and @avikivity promised to share a document about it. Apparently we both forgot about it shortly

xemul commented on utils/s3/client.cc:
```diff
@@ -113,6 +114,230 @@ future<> client::delete_object(sstring object_name) {
+        return maybe_flush();
+    }
+
+    virtual future<> flush() override {
```
> If I understand correctly, contrary to regular sinks, you should not call flush() until the very end.
> Doing otherwise will cause the sink to be terminated and you won't be able to write more into it.

Correct

> Or maybe I misunderstood the interface and you are never supposed to call flush() in the middle in general?

There's no such requirement, so any writer is free to flush as often as it wants. And s3 sink will get broken after this, I didn't put any fuse against that.

Finalization of upload happening in flush() is not to do it in close(), because it apparently may fail, and failing close() was learned to be bad (@bhalevy must have scars left from those battles)

I planned to fix it together with #13019. Every flush would finalize temporary object and would then copy that object at the tail of the target. Provided current writers only flush once, this will kill two birds with one stone -- it will make flush() callable multiple times and won't extra hurt performance on real workload.

xemul commented on utils/s3/client.cc:
```diff
@@ -338,6 +351,105 @@ data_sink client::make_upload_sink(sstring object_name) {
+    }
+
+    virtual future<size_t> read_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc) override {
+        auto buf = co_await _client->get_object_contiguous(_object_name, range{ pos, utils::iovec_len(iov) });
```
No, iovec is not contiguous buffer by nature, this is what it was invented for. But the read_dma(iovec) is supposed to read contiguous buffer from file starting from `pos` and spanning as many bytes as it fits into the iovec. In other words -- after (or while) reading, the contiguous-in-file buffer is scattered into non-contiguous-in-memory buffer pointed to by iovec.

I hope this answers your question

denesb replied:
Yes, specifically I wanted to see whether the iovec can have holes in the middle, but it sounds like this is not how it is intended to be used. Instead it facilitates reading into fragmented buffers.

xemul commented on utils/s3/client.cc:
```diff
@@ -338,6 +351,105 @@ data_sink client::make_upload_sink(sstring object_name) {
     return data_sink(std::make_unique<upload_sink>(shared_from_this(), std::move(object_name)));
 }
 
+class client::readable_file : public file_impl {
```
It's not in fact read-_only_. It's created by sstable code with flags::rw with no errors and pretends to be writable too. The only reason why it doesn't throw after it is because sstables code doesn't write directly into it, but instead creates output_stream for writing and all those places are patched to create stream via storage virtual call. I tried to avoid these hacks with scylladb/seastar#1419

raphaelsc commented on test.py:
```diff
@@ -1246,6 +1247,11 @@ async def reap(done, pending, signaled):
+    ms = MinioServer(options.tmpdir, LogPrefixAdapter(logging.getLogger('minio'), {'prefix': 'minio'}))
+    await ms.start()
+    TestSuite.artifacts.add_exit_artifact(None, ms.stop)
+
```
By the way, I'd recommend starting MinioServer only if we're running tests that will stress it.

For example, I don't want `./test.py row_cache_stress_test` to start MinioServer.

raphaelsc replied:
Yeah, that's what came to my mind.

When extending test.py for compaction groups, I introduced this param in boost/suite.yaml:

```
# Enable compaction groups on tests except on a few, listed below
all_can_run_compaction_groups_except:
    - exceptions_optimized_test
    - rate_limiter_test
    - exceptions_fallback_test
```

so we could have a param for tests that need s3 server and the suite would start minio once it first runs a test that require it. but if it's too complicated, we can perhaps postpone this improvement.

raphaelsc commented on utils/s3/client.cc:
```diff
@@ -0,0 +1,120 @@
+    auto req = http::request::make("GET", _host, object_name);
+    http::reply::status_type expected = http::reply::status_type::ok;
+    if (range) {
+        req._headers["Range"] = format("bytes={}-{}", range->off, range->off + range->len - 1);
```
Worth dumping the byte range in the trace log, so to help with debugging. For example, HTTP server might ignore the range header if there's anything wrong with it and return 200 response type on completion.

xemul replied:
> Worth dumping the byte range in the trace log

You mean the header itself, don't you? Not the actual bytes from the body :sweat_smile:

xemul commented on test/object_store/run:
```diff
@@ -0,0 +1,79 @@
+conn.execute("INSERT INTO test_ks.test_cf ( name, value ) VALUES ('1', 'one');")
+conn.execute("INSERT INTO test_ks.test_cf ( name, value ) VALUES ('2', 'two');")
+res = conn.execute("SELECT * FROM test_ks.test_cf;")
+
```
There's `test/cql-pytest/test_keyspace.py::test_storage_options_alter_type` for that already

xemul commented:
updated:
- rebased (to pick rapidxml-devel from toolchain and pre-merged PRs)
- extended object_storage test to drop the table at the end
- fixed previous review comments
- extended PR description

raphaelsc commented on sstables/sstables.cc:
```diff
@@ -141,6 +141,36 @@ future<> sstable::rename_new_sstable_component_file(sstring from_name, sstring t
+    virtual void change_dir_for_test(sstring nd) override {
+        dir = std::move(nd);
+    }
+
```
@xemul out of curiosity, why can't we reuse change_state() to achieve change of directory of tests?

xemul replied:
Because change_state() moves the whole sstable into new location. The change_dir_for_tests just forces the sstable to point at another location whatever files are there. I cannot yet wrap my head why sstable_test.cc::do_write_sst() need this hackery (it's the only user), but probably it won't like if files migrate between directories :)

raphaelsc replied:
It's indeed some sinister hackery to just rewrite some of the sstable components elsewhere :-) 

Anyway, we can revisit it later. Not much work to get rid of it.

xemul commented:
update:
- randomize s3 object names in unit test so that test case instances running together don't step on each other

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/244/

xemul commented:
`unknown location(0): fatal error: in "test_client_put_get_object": std::system_error: Connection refused`

all over the tests

xemul commented:
update:
- run `mc` in its own explicit config dir (it fails to create ~/.mc inside dbuild container with EPERM)

xemul commented:
update:
- remove adding minio to Dockerfile (it's in install-dependencies already)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/254/

avikivity commented on utils/s3/client.cc:
```diff
@@ -0,0 +1,123 @@
+        auto out = std::move(out_);
+        co_await out.write(buf.get(), buf.size());
+        co_await out.flush();
+        co_await out.close();
```
The close() needs protection in case of an exception, no?

xemul replied:
Hm... I thought close() shouldn't throw not to produce more scars to Benny. What am I missing?

avikivity replied:
I meant that close() should be called even if one of the previous calls throws. Totally against allowing close() to throw, we need to preserve @bhalevy sanity.

avikivity commented on utils/s3/client.cc:
```diff
@@ -0,0 +1,123 @@
+            co_await out.write(buf.get(), buf.size());
+        }
+        co_await out.flush();
+        co_await out.close();
```
exception handling for close()

avikivity commented on sstables/sstables.hh:
```diff
@@ -498,6 +498,7 @@ public:
         virtual future<> wipe(const sstable& sst) noexcept = 0;
         virtual future<file> open_component(const sstable& sst, component_type type, open_flags flags, file_open_options options, bool check_integrity) = 0;
         virtual future<data_sink> make_component_sink(sstable& sst, component_type type, io_priority_class pc) = 0;
+        virtual future<data_sink> make_component_sink(sstable& sst, component_type type, open_flags oflags, file_output_stream_options options) = 0;
```
These two functions aren't identical, so they should have different names. Overloading a name should be done only if the operation is the same.

avikivity commented on docs/dev/system_keyspace.md:
```diff
@@ -356,4 +356,25 @@ CREATE TABLE system.clients (
+
+Schema:
+~~~
+CREATE TABLE system.sstables_registry (
```
I think "system.sstables" is nicer, no need to be overly verbose here.

avikivity commented:
Excellent!

xemul commented:
updated:
- use test's hosts leasing to generate minio server address
- close http output streams on exceptions
- renamed "sstables_registry" table into "sstables"
- added utils/s3/client.cc to cmakefile
- de-overloaded make_component_sink methods

xemul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/266/

#13211

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/273/

xemul commented:
upd:
- conflicted with system.topology patches, rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/395/

raphaelsc commented:
@scylladb/scylla-maint I suggest we merge this series. The approach is already agreed on. Also it's experimental. Subsequent improvements can be done incrementally.

xemul commented:
> > @scylladb/scylla-maint I suggest we merge this series. The approach is already agreed on. Also it's experimental. Subsequent improvements can be done incrementally.
> 
> @xemul it's ready to be merged right?

It is now

upd:
- rebased (had conflicts)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/501/

xemul commented:
upd:
- rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/545/
